### PR TITLE
Ports `text_replace` tests to `partiql-tests` and adds null/missing/mistyped argument tests

### LINKED ISSUE #78: Add `TEXT_REPLACE` tests
Breaking apart https://github.com/partiql/partiql-tests/issues/72 to track separately.

### PR DESCRIPTION
**Pull request by alancai98**
Closes #78.

Ports the `text_replace` tests to `partiql-tests`. Also adds null/missing propagation tests and mistyped argument tests.

These tests are ported under the "extended" directory because the name isn't part of any SQL spec I could find. It appears most other SQL databases (e.g. postgresql, mysql, sqlite, oracle) use the name `REPLACE`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by vgapeyev** [comment]
> These tests are ported under the "extended" directory because the name isn't part of any SQL spec I could find. It appears most other SQL databases (e.g. postgresql, mysql, sqlite, oracle) use the name REPLACE.

The extended directory sounds right to me, since this function was mimicking Postres' non-standard `replace`.  The same name could not be used because `REPLACE` keyword is already used in PartiQL DML. (But there is nothing special about the `text_replace` choice itself.)

**Review by vgapeyev — COMMENTED** [review]
This is primarily all right, but let's see if anything transpires from a couple questions I raised.

**Comment by alancai98** [comment]
> The extended directory sounds right to me, since this function was mimicking Postres' non-standard `replace`.  The same name could not be used because `REPLACE` keyword is already used in PartiQL DML. (But there is nothing special about the `text_replace` choice itself.)

I see. Thanks for the context. I'll keep it in the "extended" directory then.

File: partiql-tests-data-extended/eval/primitives/functions/text_replace.ion [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    }
+  },
+  {
+    name:"text_replace MISSING third arg",
+    statement:"text_replace('foo', 'bar', MISSING)",
+    assert:{
+      result:EvaluationSuccess,
+      evalMode:[EvalModeCoerce, EvalModeError],
+      output:$missing::null
+    }
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-tests-data-extended/eval/primitives/functions/text_replace.ion:172]: > It seems to me the spec (Ch 7) is clear that NULL and MISSING should propagate through function calls, except for an explicit selection of several special standard functions where they do not (IS NULL, IS MISSING, AND, ...). Or am I overly optimistic in reading that?

That's correct that except for certain function calls/operators (like the `IS [NOT] NULL/MISSING` that you had mentioned and also `NULLIF` and `COALESCE`), they should propagate NULL and MISSING.

> So, in general, adding tests like these for every function would be noisy and perhaps unnecessary?
But seeing that there are no tests like these for another multi-argument function in the "extended" tests already, I guess it makes sense to add these, with understanding there was nothing special about text_replace being chosen for this "public service".

Hmm, I agree that these null/missing propagation tests could be somewhat repetitive to add for every function. I think it's still important for the conformance tests to ensure the functions properly propagate null and missing. I haven't found a less repetitive way to test for null/missing propagation than adding explicit tests though.

FYI there are some more function tests in the non-extended directory that do test for null/missing propagation. E.g. for the `MOD` function:

https://github.com/partiql/partiql-tests/blob/95e16c5018ae384bc010c232c43215928823e05a/partiql-tests-data/eval/primitives/functions/mod.ion#L98-L109

The "extended" directory may be incomplete as those other function tests were ported from `partiql-lang-kotlin`, which may not have contained null/missing propagation tests.

File: partiql-tests-data-extended/eval/primitives/functions/text_replace.ion [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  },
+  {
+    name:"text_replace NULL and MISSING arg",
+    statement:"text_replace(NULL, 'foo', MISSING)",
+    assert:{
+      result:EvaluationSuccess,
+      evalMode:[EvalModeCoerce, EvalModeError],
+      output:$missing::null
+    }
```
[the remark this answers is not in the document]
Comment (alancai98) [reply partiql-tests-data-extended/eval/primitives/functions/text_replace.ion:181]: `MISSING` over `NULL` propagation isn't discussed in the PartiQL spec. I had previously created an issue to track this: https://github.com/partiql/partiql-spec/issues/35. I discussed with Almann about this previously and we determined that `MISSING` should "win over" `NULL`. Perhaps I forgot to followup on the issue. There are some existing tests for this behavior in the non-extended directory. E.g. https://github.com/partiql/partiql-tests/blob/95e16c5018ae384bc010c232c43215928823e05a/partiql-tests-data/eval/primitives/functions/substring.ion#L1178-L1189

`partiql-lang-kotlin` in permissive mode follows this behavior.

> I guess a rationale could be that since MISSING is the permissive mode's stand-in for an error and the error would "win" over NULL in the strict mode?

That's the way I had generally viewed why `MISSING` propagates over `NULL`.

> would this new principle force us to reconsider other NULL vs MISSING design decisions?

I'm not aware of any contradictions or confusing behavior due to this decision.

**Review by vgapeyev — APPROVED** [review]
Thanks for the discussions; this is good to go.
